controllers/keystoneendpoint_controller.go

```diff
@@ -0,0 +1,422 @@
+			err.Error()))
+		return ctrl.Result{}, err
+	}
+
+	if !keystoneAPI.IsReady() {
+		instance.Status.Conditions.Set(condition.FalseCondition(
+			keystonev1.KeystoneAPIReadyCondition,
+			condition.RequestedReason,
+			condition.SeverityInfo,
+			keystonev1.KeystoneAPIReadyWaitingMessage))
+		util.LogForObject(helper, "KeystoneAPI not yet ready!", instance)
+
+		return ctrl.Result{RequeueAfter: time.Second * 5}, nil
+	}
+	instance.Status.Conditions.MarkTrue(keystonev1.KeystoneAPIReadyCondition, keystonev1.KeystoneAPIReadyMessage)
+
+	//
+	// get admin authentication OpenStack
+	//
+	os, ctrlResult, err := keystonev1.GetAdminServiceClient(
+		ctx,
+		helper,
+		keystoneAPI,
+	)
+	if err != nil {
+		instance.Status.Conditions.Set(condition.FalseCondition(
+			keystonev1.AdminServiceClientReadyCondition,
+			condition.ErrorReason,
+			condition.SeverityWarning,
+			keystonev1.AdminServiceClientReadyErrorMessage,
+			err.Error()))
+		return ctrl.Result{}, err
+	}
+	if (ctrlResult != ctrl.Result{}) {
+		instance.Status.Conditions.Set(condition.FalseCondition(
+			keystonev1.AdminServiceClientReadyCondition,
+			condition.RequestedReason,
+			condition.SeverityInfo,
+			keystonev1.AdminServiceClientReadyWaitingMessage))
+		return ctrlResult, nil
+	}
+	instance.Status.Conditions.MarkTrue(keystonev1.AdminServiceClientReadyCondition, keystonev1.AdminServiceClientReadyMessage)
+
+	// update status to save current conditions to object before sub-reconcilation rules start
+	if err := r.Status().Update(ctx, instance); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Handle endpoint delete
+	if !instance.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, instance, helper, os)
+	}
+
+	// Handle non-deleted clusters
+	return r.reconcileNormal(ctx, instance, helper, os)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *KeystoneEndpointReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&keystonev1.KeystoneEndpoint{}).
+		Complete(r)
+}
+
+func (r *KeystoneEndpointReconciler) reconcileDelete(
+	ctx context.Context,
+	instance *keystonev1.KeystoneEndpoint,
+	helper *helper.Helper,
+	os *openstack.OpenStack,
+) (ctrl.Result, error) {
+	util.LogForObject(helper, "Reconciling Endpoint delete", instance)
+
+	// only cleanup the service if there are EndpointIDs references in the
+	// object status
+	if len(instance.Status.EndpointIDs) > 0 {
+		// Delete Endpoints
+		for endpointType := range instance.Spec.Endpoints {
+			// get the gopher availability mapping for the endpointInterface
+			availability, err := openstack.GetAvailability(endpointType)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+
+			err = os.DeleteEndpoint(
+				r.Log,
+				openstack.Endpoint{
+					Name:         instance.Spec.ServiceName,
+					ServiceID:    instance.Status.ServiceID,
+					Availability: availability,
+				},
+			)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+		}
```

Comment:
Is it possible that `instance.Spec.Endpoints` might currently have more endpoint keys than `instance.Status.EndpointIDs` such that then `os.DeleteEndpoint` might be called for an endpoint that doesn't actually exist on the backend?  Maybe it doesn't ultimately matter.  Just curious.

Reply:
`allEndpints` in https://github.com/openstack-k8s-operators/lib-common/blob/master/modules/openstack/endpoint.go#L99 for that type should be empty and `os.DeleteEndpoint` won't call an actual delete. But I need to validate.

Reply:
If it is free to call delete on a not created endpoint then I think we can drop the `len(instance.Status.EndpointIDs) > 0` condition and the `else` branch too.

Reply:
yes, validated what I said and will add a commit to remove the conditionals as Gibi said above:

in Spec
~~~
# oc get keystoneendpoint placement -o json | jq .spec.endpoints
{
  "admin": "http://placement-admin-openstack.apps-crc.testing",
  "internal": "http://placement-internal-openstack.apps-crc.testing",
  "public": "http://placement-public-openstack.apps-crc.testing"
}
~~~

in keystone
~~~
# openstack endpoint list --service placement
+----------------------------------+-----------+--------------+--------------+---------+-----------+------------------------------------------------------+
| ID                               | Region    | Service Name | Service Type | Enabled | Interface | URL                                                  |
+----------------------------------+-----------+--------------+--------------+---------+-----------+------------------------------------------------------+
| 51d159e2b44e4fcd8c6a2a12348f8aa0 | regionOne | placement    | placement    | True    | admin     | http://placement-admin-openstack.apps-crc.testing    |
| 91fd568ccd1c43019a001c48c03904b0 | regionOne | placement    | placement    | True    | internal  | http://placement-internal-openstack.apps-crc.testing |
+----------------------------------+-----------+--------------+--------------+---------+-----------+------------------------------------------------------+
~~~

=> delete works fine

Reply:
done with 396862a08dea20cc594c19148a6bbb869bae79a5